plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesPanel.java

```diff
@@ -0,0 +1,392 @@
+    return bucketComposite;
+  }
+
+  /**
+   * Validates a checkbox and text field as follows:
+   * <ol>
+   * <li>if the checkbox is unselected -> valid
+   * <li>if the checkbox is selected -> the result is determined by the provided <code>validator</code> used
+   * on the value of the text field
+   * </ol>
+   *
+   */
+  private static class OverrideValidator extends MultiValidator {
+
+    private ISWTObservableValue selectionObservable;
+    private ISWTObservableValue textObservable;
+    private IValidator validator;
+
+    /**
+     * @param selection must be an observable for a checkbox, i.e. a {@link Button} with {@link SWT#CHECK} style
+     * @param text must be an observable for a {@link Text}
+     * @param validator must be a validator for String values, will be applied to <code>text.getValue()</code>
+     */
+    public OverrideValidator(ISWTObservableValue selection, ISWTObservableValue text, IValidator validator) {
+      super(selection.getRealm());
+      Preconditions.checkArgument(text.getWidget() instanceof Text,
+                                  "text is an observable for {0}, should be for {1}",
+                                  text.getWidget().getClass().getName(),
+                                  Text.class.getName());
+      Preconditions.checkArgument(selection.getWidget() instanceof Button,
+                                  "selection is an observable for {0}, should be for {1}",
+                                  selection.getWidget().getClass().getName(),
+                                  Button.class.getName());
+      Preconditions.checkArgument((selection.getWidget().getStyle() & SWT.CHECK) != 0,
+          "selection must be an observable for a checkbox");
+      this.selectionObservable = selection;
+      this.textObservable = text;
+      this.validator = validator;
+    }
+
+    @Override
+    protected IStatus validate() {
+      if (Boolean.FALSE.equals(selectionObservable.getValue())) {
+        return ValidationStatus.ok();
+      }
+      return validator.validate(textObservable.getValue());
+    }
+
+    @Override
+    public IObservableList getTargets() {
+      /**
+       * BUGFIX: https://bugs.eclipse.org/bugs/show_bug.cgi?id=312785
+       */
+      if( isDisposed() ) {
+        return Observables.emptyObservableList();
+      }
+      return super.getTargets();
+    }
+  }
+
+  public DataBindingContext getDataBindingContext() {
+    return bindingContext;
+  }
+
+  public void resetToDefaults() {
+    model.resetToDefaults();
+    bindingContext.updateTargets();
+  }
+
+  @Override
+  public void dispose() {
+    if (bindingContext != null) {
+      bindingContext.dispose();
+    }
+    if (observables != null) {
+      observables.dispose();
+    }
+    super.dispose();
+  }
+
+  private void handleExpansionStateChanged(final ExpansionEvent e) {
+    expansionHandler.handleExpansionEvent(e);
+  }
+
+  public static interface AdvancedSectionExpansionHandler {
```

Comment:
<joke>considering how much effort was to make resizing work properly (?), I don't think we should add more collapsible sections</joke>
I'll rename it to `SectionExpansionHandler`.